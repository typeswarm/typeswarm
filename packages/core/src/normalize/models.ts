import {
    ComposeSpecification,
    DefinitionsService,
    DefinitionsDeployment,
} from '../compose-spec';

//https://stackoverflow.com/questions/51465182/typescript-remove-index-signature-using-mapped-types
//https://github.com/microsoft/TypeScript/issues/25987#issuecomment-870515762
export type KnownKeys<T> = keyof {
    [K in keyof T as string extends K ? never : number extends K ? never : K]: never
  }
  

export interface StrictPortMapping {
    mode?: string;
    target?: number;
    published?: number;
    protocol?: string;

    [k: string]: unknown; //Left for backwards compatibility
}

export interface Dictionary {
    [k: string]: string | number | boolean | null;
}

export interface StrictConfig {
    source?: string;
    target?: string;
    uid?: string;
    gid?: string;
    mode?: number;

    [k: string]: unknown; //Left for backwards compatibility
}

export interface StrictVolume {
    type: string;
    source?: string;
    target?: string;
    read_only?: boolean;
    consistency?: string;
    bind?: {
        propagation?: string;
    };
    volume?: {
        nocopy?: boolean;
    };
    tmpfs?: {
        size?: number;
    };

    [k: string]: unknown; //Left for backwards compatibility
}

export interface StrictNetwork {
    aliases?: string[];
    ipv4_address?: string;
    ipv6_address?: string;
    link_local_ips?: string[];
    priority?: number;

    [k: string]: unknown; //Left for backwards compatibility
}

export type StrictDeployment = Omit<
    Pick<DefinitionsDeployment, KnownKeys<DefinitionsDeployment>>,
    'labels'
> & {
    labels?: Dictionary;
};

export type StrictService = Omit<
    Pick<DefinitionsService, KnownKeys<DefinitionsService>>,
    | 'environment'
    | 'deploy'
    | 'labels'
    | 'ports'
    | 'volumes'
    | 'configs'
    | 'secrets'
    | 'networks'
> & {
    ports?: StrictPortMapping[];
    environment?: Dictionary;
    deploy?: StrictDeployment;
    labels?: Dictionary;
    volumes?: StrictVolume[];
    configs?: StrictConfig[];
    secrets?: StrictConfig[];
    networks?: {
        [k: string]: StrictNetwork | null;
    };
};

export type StrictServicesDict = {
    [k: string]: StrictService;
};

export type StrictSpecification = Omit<
    Pick<ComposeSpecification, KnownKeys<ComposeSpecification>>,
    'services'
> & {
    services?: {
        [k: string]: StrictService;
    };
};
