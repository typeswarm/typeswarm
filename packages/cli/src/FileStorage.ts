import { readFile, writeFile } from 'fs/promises';
import { injectable } from 'inversify';

export interface IFileStorage {
    read(fileName: string): Promise<string>;
    write(fileName: string, contents: string): Promise<void>;
}

@injectable()
export class FileStorageImpl_Real implements IFileStorage {
    read(fileName: string): Promise<string> {
        return readFile(fileName, 'utf8');
    }
    write(fileName: string, contents: string): Promise<void> {
        return writeFile(fileName, contents, 'utf8');
    }
}

@injectable()
export class FileStorageImpl_Mock implements IFileStorage {
    public files: Record<string, string> = {};

    read(fileName: string): Promise<string> {
        if (!this.files[fileName]) {
            throw new Error(`File not found: ${fileName}`);
        }
        return Promise.resolve(this.files[fileName]);
    }

    write(fileName: string, contents: string): Promise<void> {
        this.files[fileName] = contents;
        return Promise.resolve();
    }
}
