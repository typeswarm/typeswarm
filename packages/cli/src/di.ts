export const Types = {
    Logger: Symbol('Logger'),
    EntitiesProcessor: Symbol('EntitiesProcessor'),
    ComposeBuilder: Symbol('ComposeBuilder'),
    CLIApplication: Symbol('CLIApplication'),
    IFileStorage: Symbol('IFileStorage'),
};
