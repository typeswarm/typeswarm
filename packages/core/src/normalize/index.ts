import { DefinitionsService } from '../compose-spec';
import { StrictService } from './models';

export * from './models';

export * from './parseConfigs';
export * from './parseListOrDict';
export * from './parsePorts';
export * from './parseVolumes';

export * from './parseService';
export * from './parseSpecification';
