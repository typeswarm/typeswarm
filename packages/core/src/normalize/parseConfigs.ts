import { DefinitionsService } from '../compose-spec';
import { StrictConfig } from './models';

export function parseConfigs(
    configs: DefinitionsService['configs'] | DefinitionsService['secrets']
): StrictConfig[] | undefined {
    if (!configs) {
        return undefined;
    }
    return configs.map(parseConfig);
}

function parseConfig(config: StrictConfig | string): StrictConfig {
    if (typeof config === 'string') {
        return {
            source: config,
            target: `/${config}`,
        };
    }
    return config;
}
