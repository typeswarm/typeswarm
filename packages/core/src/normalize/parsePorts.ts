import { StrictPortMapping } from './models';
import { DefinitionsService } from '../compose-spec';
import { range } from 'lodash';

export function parsePorts(
    ports: DefinitionsService['ports']
): StrictPortMapping[] | undefined {
    if (!ports) {
        return undefined;
    }
    if (!ports.length) {
        return [];
    }

    return ports.flatMap(parsePort);
}

function parsePort(
    port: StrictPortMapping | number | string
): StrictPortMapping[] {
    if (typeof port === 'number') {
        return [
            {
                target: port,
            },
        ];
    }
    if (typeof port === 'string') {
        return parsePortStr(port);
    }
    return [port];
}

function parsePortStr(mappingWithProto: string): StrictPortMapping[] {
    const [mapping, proto] = mappingWithProto.split('/');
    const groups = mapping.match(/((\d+(-\d+)?):)?(\d+(-\d+)?)$/);
    if (!groups) {
        throw new SyntaxError(
            `Could not parse port mapping ${mappingWithProto}`
        );
    }
    const publishedRange =
        groups[2] === undefined ? undefined : parseRange(groups[2]);
    const targetRange = parseRange(groups[4]);

    if (targetRange && publishedRange) {
        if (targetRange.length !== publishedRange.length) {
            throw new RangeError(`Unequal port ranges ${mappingWithProto}`);
        }
    }

    return targetRange.map((targetPort, i) => {
        const mapping: StrictPortMapping = {};
        mapping.target = targetPort;
        if (publishedRange !== undefined) {
            mapping.published = publishedRange[i];
        }
        if (proto !== undefined) {
            mapping.protocol = proto;
        }
        return mapping;
    });
}

function parseRange(portOrRange: string): number[] {
    const groups = portOrRange.match(/^(\d+)(-(\d+))?$/);
    if (!groups) {
        throw new SyntaxError(`Could not parse port range ${portOrRange}`);
    }
    const start = parseInt(groups[1]);
    const end = groups[3] === undefined ? start : parseInt(groups[3]);

    if (end < start) {
        throw new RangeError(`Invalid ports range ${portOrRange}`);
    }
    if (end - start > 1000) {
        throw new RangeError(`Too big ports range ${portOrRange}`);
    }
    return range(start, end + 1);
}
