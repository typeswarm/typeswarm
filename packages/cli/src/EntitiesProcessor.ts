import 'reflect-metadata';

import {
    DefinitionsConfig,
    DefinitionsSecret,
    PropertiesConfigs,
} from '@typeswarm/core';
import {
    StrictServicesDict,
    StrictSpecification,
} from '@typeswarm/core/lib/normalize';
import { getHash } from '@typeswarm/core/lib/utils';

import { inject, injectable } from 'inversify';
import { basename, join, relative } from 'path';
import { Logger } from 'tslog';
import { Types } from './di';
import { IFileStorage } from './FileStorage';

@injectable()
export class EntitiesProcessor {
    constructor(
        @inject(Types.Logger)
        private logger: Logger,
        @inject(Types.IFileStorage)
        private fileStorage: IFileStorage
    ) {}

    private async processSingleEntity(
        name: string,
        entity: DefinitionsConfig | DefinitionsSecret,
        directory: string
    ): Promise<{ rotatedEntityName: string; rotatedEntityFile: string }> {
        if (!entity.file && !entity.data) {
            throw new Error(
                '.file or .data is missing in ' + JSON.stringify(entity)
            );
        }
        //TODO: support raw binary data
        const contents = entity.data
            ? entity.data
            : await this.fileStorage.read(entity.file as string);

        const hash = getHash(contents);
        const rotatedEntityFileBaseName = entity.data
            ? `${hash}.${name}.config` //TODO: slugify name
            : `${hash}.${basename(entity.file as string)}`;

        const rotatedEntityFile = join(directory, rotatedEntityFileBaseName);
        this.logger.info('Write file', rotatedEntityFile);
        //TODO: support raw binary data
        await this.fileStorage.write(rotatedEntityFile, contents);

        const rotatedEntityName = `${name}_${hash}`;
        return {
            rotatedEntityName,
            rotatedEntityFile,
        };
    }

    async processEntities(
        entity: 'secrets' | 'configs',
        spec: StrictSpecification,
        directory: string
    ): Promise<StrictSpecification> {
        const {
            //secrets or configs
            [entity]: entitiesMap = {},
            services = {},
        } = spec;
        const rotatedEntities: PropertiesConfigs = {};
        const rotatedServices: StrictServicesDict = {};

        const rotationMap: { [configName: string]: string } = {};

        for (const [name, config] of Object.entries(entitiesMap)) {
            if (
                config.name ||
                config.external ||
                !(config.file || config.data)
            ) {
                rotatedEntities[name] = config;
                continue;
            }
            const {
                rotatedEntityName: rotatedEntityName,
                rotatedEntityFile: rotatedEntityFile,
            } = await this.processSingleEntity(name, config, directory);
            rotatedEntities[rotatedEntityName] = {
                ...config,
                file: relative(directory, rotatedEntityFile),
            };
            delete rotatedEntities[rotatedEntityName].data;
            rotationMap[name] = rotatedEntityName;
        }

        for (const [name, service] of Object.entries(services)) {
            const {
                //service.configs or service.secrets
                [entity]: entityReferences = [],
            } = service;
            const rotatedEntityReferences: typeof entityReferences = [];
            for (const entityRef of entityReferences) {
                if (
                    typeof entityRef === 'object' &&
                    entityRef.source &&
                    rotationMap[entityRef.source]
                ) {
                    rotatedEntityReferences.push({
                        ...entityRef,
                        source: rotationMap[entityRef.source],
                    });
                } else {
                    rotatedEntityReferences.push(entityRef);
                }
            }
            rotatedServices[name] = {
                ...service,
                [entity]: rotatedEntityReferences,
            };
            if (rotatedEntityReferences.length === 0) {
                delete rotatedServices[name][entity];
            }
        }
        return {
            ...spec,
            [entity]: rotatedEntities,
            services: rotatedServices,
        };
    }
}
