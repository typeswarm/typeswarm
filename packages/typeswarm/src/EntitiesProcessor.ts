import fs from 'fs';
import { inject, injectable } from 'inversify';
import { basename, join, relative } from 'path';
import { Logger } from 'tslog';
import {
    DefinitionsConfig,
    DefinitionsSecret,
    PropertiesConfigs,
} from './compose-spec';
import { Types } from './di';
import { StrictServicesDict, StrictSpecification } from './normalize';
import { getHash } from './utils';

@injectable()
export class EntitiesProcessor {
    constructor(
        @inject(Types.Logger)
        private logger: Logger
    ) {}

    private async processSingleEntity(
        name: string,
        entity: DefinitionsConfig | DefinitionsSecret,
        directory: string
    ): Promise<{ rotatedEntityName: string; rotatedEntityFile: string }> {
        if (!entity.file && !entity.data) {
            throw new Error(
                '.file or .data is missing in ' + JSON.stringify(entity)
            );
        }
        //TODO: support raw binary data
        const contents = entity.data
            ? entity.data
            : await fs.promises.readFile(entity.file as string, 'utf8');

        const hash = getHash(contents);
        const rotatedEntityFileBaseName = entity.data
            ? `${hash}.${name}.config` //TODO: slugify name
            : `${hash}.${basename(entity.file as string)}`;

        const rotatedEntityFile = join(directory, rotatedEntityFileBaseName);
        this.logger.info('Write file', rotatedEntityFile);
        //TODO: support raw binary data
        await fs.promises.writeFile(rotatedEntityFile, contents, 'utf8');

        const rotatedEntityName = `${name}_${hash}`;
        return {
            rotatedEntityName,
            rotatedEntityFile,
        };
    }

    async processEntities(
        entity: 'secrets' | 'configs',
        spec: StrictSpecification,
        directory: string
    ): Promise<StrictSpecification> {
        const {
            //secrets or configs
            [entity]: entitiesMap = {},
            services = {},
        } = spec;
        const rotatedEntities: PropertiesConfigs = {};
        const rotatedServices: StrictServicesDict = {};

        const rotationMap: { [configName: string]: string } = {};

        for (const [name, config] of Object.entries(entitiesMap)) {
            if (
                config.name ||
                config.external ||
                !(config.file || config.data)
            ) {
                rotatedEntities[name] = config;
                continue;
            }
            const {
                rotatedEntityName: rotatedEntityName,
                rotatedEntityFile: rotatedEntityFile,
            } = await this.processSingleEntity(name, config, directory);
            rotatedEntities[rotatedEntityName] = {
                ...config,
                file: relative(directory, rotatedEntityFile),
            };
            delete rotatedEntities[rotatedEntityName].data;
            rotationMap[name] = rotatedEntityName;
        }

        for (const [name, service] of Object.entries(services)) {
            const {
                //service.configs or service.secrets
                [entity]: entityReferences = [],
            } = service;
            const rotatedEntityReferences: typeof entityReferences = [];
            for (const entityRef of entityReferences) {
                if (
                    typeof entityRef === 'object' &&
                    entityRef.source &&
                    rotationMap[entityRef.source]
                ) {
                    rotatedEntityReferences.push({
                        ...entityRef,
                        source: rotationMap[entityRef.source],
                    });
                } else {
                    rotatedEntityReferences.push(entityRef);
                }
            }
            rotatedServices[name] = {
                ...service,
                [entity]: rotatedEntityReferences,
            };
            if (rotatedEntityReferences.length === 0) {
                delete rotatedServices[name][entity];
            }
        }
        return {
            ...spec,
            [entity]: rotatedEntities,
            services: rotatedServices,
        };
    }
}
