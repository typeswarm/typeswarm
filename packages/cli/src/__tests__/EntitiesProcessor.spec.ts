import { StrictService } from '@typeswarm/core/lib/normalize';
import { Container } from 'inversify';
import { Logger } from 'tslog';
import { Types } from '../di';
import { EntitiesProcessor } from '../EntitiesProcessor';
import { FileStorageImpl_Mock, IFileStorage } from '../FileStorage';

function createContainer() {
    const di = new Container();
    di.bind<EntitiesProcessor>(Types.EntitiesProcessor).to(EntitiesProcessor);
    di.bind<IFileStorage>(Types.IFileStorage)
        .to(FileStorageImpl_Mock)
        .inSingletonScope();
    di.bind<Logger>(Types.Logger).toConstantValue(
        new Logger({ displayFilePath: 'hidden' })
    );
    return di;
}

describe('EntitiesProcessor', () => {
    it('should process both configs and secrets', async () => {
        const di = createContainer();

        const entitiesProcessor = di.get<EntitiesProcessor>(
            Types.EntitiesProcessor
        );

        const specWithRotatedConfigs = await entitiesProcessor.processEntities(
            'configs',
            {
                services: {
                    website: {
                        image: 'my-image:latest',
                        configs: [
                            {
                                source: 'website-config.yaml',
                                target: '/etc/app/config.yaml',
                            },
                        ],
                        secrets: [
                            {
                                source: 'website-credentials.yaml',
                                target: '/etc/app/credentials.yaml',
                            },
                        ],
                    } as StrictService,
                },
                configs: {
                    'website-config.yaml': {
                        data: 'hello: world',
                    },
                },
                secrets: {
                    'website-credentials.yaml': {
                        data: 'apiKey: "1234567890"',
                    },
                },
            },
            'build'
        );

        const specWithRotatedSecrets = await entitiesProcessor.processEntities(
            'secrets',
            specWithRotatedConfigs,
            'build'
        );

        expect(specWithRotatedSecrets).toMatchSnapshot();
        const fileStorage = di.get<FileStorageImpl_Mock>(Types.IFileStorage);
        expect(fileStorage.files).toMatchSnapshot();
    });
});
