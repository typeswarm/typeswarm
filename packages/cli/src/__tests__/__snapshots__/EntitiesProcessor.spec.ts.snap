// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitiesProcessor should process both configs and secrets 1`] = `
Object {
  "configs": Object {
    "website-config.yaml_ba9f826f": Object {
      "file": "ba9f826f.website-config.yaml.config",
    },
  },
  "secrets": Object {
    "website-credentials.yaml_f709d843": Object {
      "file": "f709d843.website-credentials.yaml.config",
    },
  },
  "services": Object {
    "website": Object {
      "configs": Array [
        Object {
          "source": "website-config.yaml_ba9f826f",
          "target": "/etc/app/config.yaml",
        },
      ],
      "image": "my-image:latest",
      "secrets": Array [
        Object {
          "source": "website-credentials.yaml_f709d843",
          "target": "/etc/app/credentials.yaml",
        },
      ],
    },
  },
}
`;

exports[`EntitiesProcessor should process both configs and secrets 2`] = `
Object {
  "build/ba9f826f.website-config.yaml.config": "hello: world",
  "build/f709d843.website-credentials.yaml.config": "apiKey: \\"1234567890\\"",
}
`;
